src/storage/src/sink/kafka.rs

```diff
@@ -539,118 +532,88 @@ impl KafkaSinkState {
+    async fn retry_on_txn_error<'a, F, Fut, T>(&self, f: F) -> T
     where
         F: Fn(KafkaTxProducer) -> Fut,
         Fut: Future<Output = KafkaResult<T>>,
     {
-        let shutdown = Cell::new(false);
-        let self_producer = self.producer.clone();
-        let mut last_error = KafkaError::Canceled;
-        // Only actually used for retriable errors.
-        let tries = Retry::default()
-            .max_tries(usize::MAX)
-            .clamp_backoff(Duration::from_secs(60 * 10))
-            .into_retry_stream();
-        tokio::pin!(tries);
-        while tries.next().await.is_some() {
-            match f(self_producer.clone()).await {
-                Ok(result) => return Ok(result),
-                Err(KafkaError::Transaction(e)) => {
-                    // clone is a cheap: Option<Arc<..>> internally
-                    last_error = KafkaError::Transaction(e.clone());
-                    if e.txn_requires_abort() {
-                        info!("Error requiring tx abort in kafka sink: {:?}", e);
-                        let self_self_producer = self_producer.clone();
-                        // Only actually used for retriable errors.
-                        let should_shutdown = Retry::default()
-                            .clamp_backoff(Duration::from_secs(60 * 10))
-                            .retry_async(|_| async {
-                                info!("Attempting to abort kafka transaction");
-                                match self_self_producer.abort_transaction().await {
-                                    Ok(_) => Ok(false),
-                                    Err(KafkaError::Transaction(e)) if e.is_retriable() => {
-                                        Err(KafkaError::Transaction(e))
-                                    }
-                                    Err(e) => {
-                                        error!("Error aborting kafka transaction: {:?}", e);
-                                        Ok(true)
-                                    }
-                                }
-                            })
-                            .await
-                            .unwrap_or(true);
-                        shutdown.set(should_shutdown);
-                    } else if e.is_retriable() {
-                        info!("Retriable error in kafka sink: {:?}", e);
-                        continue;
-                    } else {
-                        shutdown.set(true);
+        Retry::default()
+            .clamp_backoff(BACKOFF_CLAMP)
+            .retry_async(|_| async {
+                match f(self.producer.clone()).await {
+                    Ok(result) => Ok(result),
+                    Err(KafkaError::Transaction(e)) if e.txn_requires_abort() => {
+                        info!("error requiring txn abort in kafka sink: {:?}", e);
+                        self.abort_active_txn().await;
+                        Err(KafkaError::Transaction(e))
```

Comment:
I believe we want this to be `Ok(_)`, break the loop and can return an error -- or panic, depending how aggressive we want to be.  (@bkirwi thoughts?).

I don't think it's generally safe to just retry whatever operation after we abort the transaction!

Reply:
Yeah, agree this is a bit nervewracking. This happens to give fine behaviour at the moment IIUC (`retry_on_txn_error` is only called for `init_transactions`, `commit_transaction`, and `begin_transaction`; I think the first two will return non-retriable errors if the transaction is aborted, and `begin_transaction` will do the right thing) but it's not a safe assumption in general.